library/src/main/java/org/eclipse/uprotocol/ULink.java

```diff
@@ -0,0 +1,763 @@
+        }
+    }
+
+    private static @NonNull PackageInfo getPackageInfo(@NonNull Context context) {
+        try {
+            return context.getPackageManager().getPackageInfo(context.getPackageName(),
+                    PackageManager.GET_SERVICES | PackageManager.GET_META_DATA);
+        } catch (NameNotFoundException e) {
+            throw new SecurityException(e.getMessage(), e);
+        }
+    }
+
+    private static @NonNull UEntity checkContainsEntity(@NonNull PackageInfo packageInfo, @Nullable UEntity entity) {
+        return Stream.concat(Stream.of(packageInfo.applicationInfo),
+                        (packageInfo.services != null) ? Stream.of(packageInfo.services) : Stream.empty())
+                .map(info -> {
+                    final UEntity foundEntity = getEntity(info);
+                    if (entity != null) {
+                        return entity.equals(foundEntity) ? entity : null;
+                    } else {
+                        return foundEntity;
+                    }
+                })
+                .filter(Objects::nonNull)
+                .findFirst()
+                .orElseThrow(() -> new SecurityException("Missing or not matching '" + META_DATA_ENTITY_NAME +
+                        "' and '" + META_DATA_ENTITY_VERSION + "' meta-data in manifest"));
+    }
+
+    private static @Nullable UEntity getEntity(@Nullable PackageItemInfo info) {
+        if (info != null && info.metaData != null) {
+            final String name = info.metaData.getString(META_DATA_ENTITY_NAME);
+            final int version = info.metaData.getInt(META_DATA_ENTITY_VERSION);
+            if (!isNullOrEmpty(name) && version > 0) {
+                return UEntity.newBuilder()
+                        .setName(name)
+                        .setVersionMajor(version)
+                        .build();
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Create an instance.
+     *
+     * @param context  An application {@link Context}. This should not be {@code null}. If you are passing
+     *                 {@link ContextWrapper}, make sure that its base Context is non-null as well.
+     *                 Otherwise it will throw {@link NullPointerException}.
+     * @param handler  A {@link Handler} on which callbacks should execute, or {@code null} to execute on
+     *                 the application's main thread.
+     * @param listener A {@link ServiceLifecycleListener} for monitoring the uBus lifecycle.
+     * @return A {@link ULink} instance.
+     * @throws SecurityException If the caller does not have {@value META_DATA_ENTITY_NAME} and
+     *         {@value META_DATA_ENTITY_VERSION} <code>meta-data</code> elements declared in the manifest.
+     * @see #META_DATA_ENTITY_NAME
+     * @see #META_DATA_ENTITY_VERSION
+     */
+    public static @NonNull ULink create(@NonNull Context context, @Nullable Handler handler,
```

Comment:
Can we consider composition with UTransport here? This would enable some custom transport implementations down the road.

Reply:
uLink is an actual transport implementation.